src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/ProxyRoutes.scala

```diff
@@ -0,0 +1,93 @@
+package org.broadinstitute.dsde.workbench.leonardo.api
+
+import akka.actor.ActorSystem
+import akka.http.scaladsl.Http
+import akka.http.scaladsl.model._
+import akka.http.scaladsl.model.ws._
+import akka.http.scaladsl.server.Directives._
+import akka.http.scaladsl.server.Route
+import akka.stream.Materializer
+import akka.stream.scaladsl._
+import com.typesafe.scalalogging.LazyLogging
+
+import scala.collection.immutable
+import scala.concurrent.duration._
+import scala.concurrent.{ExecutionContext, Future}
+
+/**
+  * Created by rtitle on 8/4/17.
+  */
+trait ProxyRoutes { self: LazyLogging =>
+  implicit val system: ActorSystem
+  implicit val materializer: Materializer
+  implicit val executionContext: ExecutionContext
+
+  val jupyterPort = 8000
+
+  val proxyRoutes: Route =
+    pathPrefix(Segment) { notebookId =>  // TODO notebookId not currently used
+      extractRequest { request =>
+        complete {
+          request.header[UpgradeToWebSocket] match {
+            case Some(upgrade) => handleWebSocketRequest(request, upgrade)
+            case None => handleHttpRequest(request)
+          }
+        }
+      }
+    }
+
+  private def handleHttpRequest(request: HttpRequest): Future[HttpResponse] = {
+    logger.info(s"Opening http connection to ${request.uri.authority.host.address}")
+    // TODO eventually we will lookup the outgoing address from the notebook ID
```

Comment:
a) yes it forwards the requests to the notebook server and returns the responses back to the caller.
b) I wonder if it will introduce too much latency to hit `ClusterComponent` for _every_ proxied notebook request. Maybe we could have a cache of `(google_project, cluster_name) -> IP` which the proxy could look up, and only query the DB on cache misses.